agent/proxycfg/api_gateway.go

```diff
@@ -0,0 +1,492 @@
+package proxycfg
+
+import (
+	"context"
+	"fmt"
+
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/proxycfg/internal/watch"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/proto/pbpeering"
+)
+
+var _ kindHandler = (*handlerAPIGateway)(nil)
+
+// handlerAPIGateway generates a new ConfigSnapshot in response to
+// changes related to an api-gateway.
+type handlerAPIGateway struct {
+	handlerState
+}
+
+// initialize sets up the initial watches needed based on the api-gateway registration
+func (h *handlerAPIGateway) initialize(ctx context.Context) (ConfigSnapshot, error) {
+	snap := newConfigSnapshotFromServiceInstance(h.serviceInstance, h.stateConfig)
+
+	// Watch for root changes
+	err := h.dataSources.CARoots.Notify(ctx, &structs.DCSpecificRequest{
+		Datacenter:   h.source.Datacenter,
+		QueryOptions: structs.QueryOptions{Token: h.token},
+		Source:       *h.source,
+	}, rootsWatchID, h.ch)
+	if err != nil {
+		return snap, err
+	}
+
+	// Get information about the entire service mesh.
+	err = h.dataSources.ConfigEntry.Notify(ctx, &structs.ConfigEntryQuery{
+		Kind:           structs.MeshConfig,
+		Name:           structs.MeshConfigMesh,
+		Datacenter:     h.source.Datacenter,
+		QueryOptions:   structs.QueryOptions{Token: h.token},
+		EnterpriseMeta: *structs.DefaultEnterpriseMetaInPartition(h.proxyID.PartitionOrDefault()),
+	}, meshConfigEntryID, h.ch)
+	if err != nil {
+		return snap, err
+	}
+
+	// Watch the api-gateway's config entry
+	err = h.subscribeToConfigEntry(ctx, structs.APIGateway, h.service, gatewayConfigWatchID)
+	if err != nil {
+		return snap, err
+	}
+
+	// Watch the bound-api-gateway's config entry
+	err = h.subscribeToConfigEntry(ctx, structs.BoundAPIGateway, h.service, gatewayConfigWatchID)
+	if err != nil {
+		return snap, err
+	}
+
+	snap.APIGateway.HTTPRoutes = watch.NewMap[structs.ResourceReference, *structs.HTTPRouteConfigEntry]()
+	snap.APIGateway.TCPRoutes = watch.NewMap[structs.ResourceReference, *structs.TCPRouteConfigEntry]()
+	snap.APIGateway.Certificates = watch.NewMap[structs.ResourceReference, *structs.InlineCertificateConfigEntry]()
+
+	snap.APIGateway.WatchedDiscoveryChains = make(map[UpstreamID]context.CancelFunc)
+	snap.APIGateway.DiscoveryChain = make(map[UpstreamID]*structs.CompiledDiscoveryChain)
+	snap.APIGateway.WatchedUpstreams = make(map[UpstreamID]map[string]context.CancelFunc)
+	snap.APIGateway.WatchedUpstreamEndpoints = make(map[UpstreamID]map[string]structs.CheckServiceNodes)
+	snap.APIGateway.WatchedGateways = make(map[UpstreamID]map[string]context.CancelFunc)
+	snap.APIGateway.WatchedGatewayEndpoints = make(map[UpstreamID]map[string]structs.CheckServiceNodes)
```

Comment:
Can you separate these from the rest, they appear to be handled via the delegated `handleUpdateUpstreams` call, and hence need to be initialized, but I'm not entirely sure what all of them are used for. Just so we don't drop them when they need to be set can you also add a comment that these must be set for the upstream handling?